### PR TITLE
Mini-library to measure execution durations and publish to cloudwatch

### PR DESCRIPTION
**Pull request by schatten**
Summary
---------
* Add a mini-library to measure execution durations and publish to AWS 
* Adds instrumentation to PDQ lambdas. At the end of the lambdas, will flush to cloudwatch.
* Terraform variable to allow measuring metrics

The scripts that will be used to generate load on the lambdas will be added after this is merged.

To the Reviewer, please review each commit separately. The commits are arranged for sane reviewing. 

Test Plan
---------
Individual commits have test plans. Wish we could do stacked diffs here.

### CODE REVIEW

**Review by BarrettOlson — COMMENTED** [review]
Added some questions inline. The approach mostly makes sense to me but before merge I wanted to call out that I am surprised we need to hand roll this much and wondering if there is anything here we can move to the terraform layer and pass in. Thanks!

File: hasher-matcher-actioner/hmalib/metrics/__init__.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    Because metric naming can have a very real impact on how dashboards are
+    built, use this central location to allow easier following of conventions.
+
+    Conventions:
+    - The metric name should be of the format "{prefix}.{action}_{noun}"
+    - prefixes can be dotted.
+    - action and noun both should be free of special characters
+    - if used as a timer, will be suffixed by "-count" and "-duration"
+    """
+    hma_namespace = "ThreatExchange/HMA"
```
[the remark this answers is not in the document]
Comment (schatten) [reply hasher-matcher-actioner/hmalib/metrics/__init__.py:35]: Totally! Will make the change.

File: hasher-matcher-actioner/terraform/pdq-signals/main.tf [hunk 2]
```diff
@@ -97,6 +98,11 @@ data "aws_iam_policy_document" "pdq_indexer" {
     ]
     resources = ["${aws_cloudwatch_log_group.pdq_indexer.arn}:*"]
   }
+  statement {
+    effect    = "Allow"
+    actions   = ["cloudwatch:PutMetricData"]
+    resources = ["*"]
```
[the remark this answers is not in the document]
Comment (schatten) [reply hasher-matcher-actioner/terraform/pdq-signals/main.tf:104]: There is no arn to which cloudwatch metrics are published. There is no destination.

The restriction comes in the form of which actions are allowed.

File: hasher-matcher-actioner/tests/test_metrics_reporter.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+def worker():
+    # just spend some time
+    with timer(names.pdq_hasher_lambda.download_file):
+        time.sleep(random.random()/100.0)
+
+    with timer(names.pdq_hasher_lambda.hash):
+        time.sleep(random.random()/100.0)
+
+def main():
+    reporter = AWSCloudWatchReporter(namespace="ThreatExchange/HMA-Test"),
```
[the remark this answers is not in the document]
Comment (schatten) [reply hasher-matcher-actioner/tests/test_metrics_reporter.py:25]: This is only a test. I do not expect partners to run it.

File: hasher-matcher-actioner/terraform/pdq-signals/variables.tf [hunk 4]
```diff
@@ -77,3 +77,9 @@ variable "additional_tags" {
   type        = map(string)
   default     = {}
 }
+
+variable "measure_performance" {
+  description = "Send metrics to cloudwatch. Useful for benchmarking, but can incur costs. Set to string True for this to work."
+  type        = string
```
[the remark this answers is not in the document]
Comment (schatten) [reply hasher-matcher-actioner/terraform/pdq-signals/variables.tf:83]: Makes more sense. I'll make the change.

File: hasher-matcher-actioner/hmalib/metrics/__init__.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+measure_performance : bool = os.getenv(_ENABLE_PERF_MEASUREMENTS_ENVVAR, "False") in ["True", "1"]
+
+logger = logging.getLogger(__name__)
+
+class names():
+    """
+    Not a real class, just a bag of metric names. Ignore the lowercase name if
+    you can. :)
+
+    Because metric naming can have a very real impact on how dashboards are
```
[the remark this answers is not in the document]
Comment (schatten) [reply hasher-matcher-actioner/hmalib/metrics/__init__.py:26]: I'd be fine with the namespace being part of terraform. I'll update the diff accordingly. Do not see value in allowing the individual metric names to be changed.

Do you foresee a case where names we provide will not suffice or will conflict with what the partner wants?

File: hasher-matcher-actioner/tests/test_metrics_reporter.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (BarrettOlson) [reply hasher-matcher-actioner/tests/test_metrics_reporter.py:25]: ah, my mistake. Maybe have the namespace be even more specific (like `-Test-Metric-Reporter`) also thoughts on moving this file to tests/script/ and call it something like `push_fake_metrics_from_reporter.py`?

**Review by BarrettOlson — APPROVED** [review]
LGTM (a lot of source code for a counter :P )

**Pull request merged**